packages/core/src/crawlers/crawler_commons.ts

```diff
@@ -5,25 +5,89 @@ import type { Response as GotResponse, OptionsInit } from 'got-scraping';
 import type { EnqueueLinksOptions } from '../enqueue_links/enqueue_links';
 import type { Log } from '../log';
 import type { ProxyInfo } from '../proxy_configuration';
-import type { Request } from '../request';
+import type { Request, Source } from '../request';
 import type { Session } from '../session_pool/session';
-import { type Dataset } from '../storages';
+import type { RequestQueueOperationOptions, Dataset, KeyValueStore } from '../storages';
 
 // eslint-disable-next-line @typescript-eslint/ban-types
-export interface CrawlingContext<Crawler = unknown, UserData extends Dictionary = Dictionary> extends Record<string & {}, unknown> {
-    id: string;
+export interface RestrictedCrawlingContext<UserData extends Dictionary = Dictionary> extends Record<string & {}, unknown>{
     /**
      * The original {@apilink Request} object.
      */
     request: Request<UserData>;
+
+    /**
+     * This function allows you to push data to a {@apilink Dataset} specified by name, or the one currently used by the crawler.
+     *
+     * Shortcut for `crawler.pushData()`.
+     *
+     * @param [data] Data to be pushed to the default dataset.
+     */
+    pushData(data: Parameters<Dataset['pushData']>[0], datasetIdOrName?: string): Promise<void>;
+
+    /**
+     * This function automatically finds and enqueues links from the current page, adding them to the {@apilink RequestQueue}
+     * currently used by the crawler.
+     *
+     * Optionally, the function allows you to filter the target links' URLs using an array of globs or regular expressions
+     * and override settings of the enqueued {@apilink Request} objects.
+     *
+     * Check out the [Crawl a website with relative links](https://crawlee.dev/docs/examples/crawl-relative-links) example
+     * for more details regarding its usage.
+     *
+     * **Example usage**
+     *
+     * ```ts
+     * async requestHandler({ enqueueLinks }) {
+     *     await enqueueLinks({
+     *       globs: [
+     *           'https://www.example.com/handbags/*',
+     *       ],
+     *     });
+     * },
+     * ```
+     *
+     * @param [options] All `enqueueLinks()` parameters are passed via an options object.
+     */
+    enqueueLinks: (options?: Omit<EnqueueLinksOptions, 'requestQueue'>) => Promise<unknown>;
```

Comment:
Why is this omitting the requestQueue? It'd be a breaking change (and also clash with the same definition in CrawlingContext)

Reply:
`CrawlingContext` extends this without a problem, so I don't think there's a clash. The reasoning is simply that 1) you shouldn't need to do this in the vast majority of use cases and 2) it's easier to swap in alternative implementations when there's less parameters (e.g., an implementation that delays the enqueueing and/or inspects the links being enqueued, adaptive crawler is the primary motivation)

Reply:
> It'd be a breaking change 

Since nobody uses this interface (yet) and the method is overriden in `CrawlingContext`, I don't really see how?

Reply:
Yeah, this PR is a huge lesson in communicating intent - I really should've explained the motivation in the description when I opened this. Let me do that now so that I don't explain it piece by piece in review comments.